File: src/dfs/FlattenBinaryTreeToLinkedList.java
package dfs;

import java.util.LinkedList;
import java.util.List;
import java.util.Stack;

import org.junit.Test;

import utility.TreeNode;

/**
 * 
 */

public class FlattenBinaryTreeToLinkedList
{
    public void flatten( TreeNode root )
    {
    	if ( root == null )
    	{
    		return;
    	}
    	
    	Stack<TreeNode> preorderStack = new Stack<>();
    	preorderStack.push( root );
    	TreeNode sentinelHead = new TreeNode( 0 );
    	TreeNode sentinelTail = sentinelHead;
    	while ( !preorderStack.isEmpty() )
    	{
    		TreeNode stackTop = preorderStack.pop();
    		if ( stackTop.right != null )
    		{
    			preorderStack.push( stackTop.right );
    		}
    		if ( stackTop.left != null )
    		{
    			preorderStack.push( stackTop.left );
    		}
    		sentinelTail.right = stackTop;
    		sentinelTail = stackTop;

    		stackTop.left = null;
    		stackTop.right = null;
    	}   	
    }

    @Test
    public void test()
    {
    	TreeNode node0 = new TreeNode( 0 );
    	TreeNode node1 = new TreeNode( 1 );
    	TreeNode node2 = new TreeNode( 2 );
    	TreeNode node3 = new TreeNode( 3 );
    	TreeNode node4 = new TreeNode( 4 );
    	TreeNode node5 = new TreeNode( 5 );
    	node0.left = node1;
    	node0.right = node4;
    	node1.left = node2;
    	node1.right = node3;
    	node4.right = node5;
    	flatten( node0 );
    }
}

File: Python/FlattenBinaryTreeToLinkedList.py
# Definition for a binary tree node.
import unittest

# Read about enumerate in python

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class FlattenBinaryTreeToLinkedList(unittest.TestCase):
    def flatten(self, root: TreeNode) -> None:
        self.flattenRecursive(root)
        print('result')

    def flattenRecursive(self, root: TreeNode) -> (TreeNode, TreeNode):
        if root is None:
            return None, None
        elif root.left is None and root.right is None:
            return root, root
        elif root.left is None and root.right is not None:
            rightListHead, rightListTail = self.flattenRecursive(root.right)
            root.right = rightListHead
            root.left = None
            return root, rightListTail
        elif root.left is not None and root.right is None:
            leftListHead, leftListTail = self.flattenRecursive(root.left)
            root.right = leftListHead
            root.left = None
            return root, leftListTail
        else:
            leftListHead, leftListTail = self.flattenRecursive(root.left)
            rightListHead, rightListTail = self.flattenRecursive(root.right)

            root.left = None
            root.right = leftListHead
            leftListTail.right = rightListHead

            return root, rightListTail

    def test_ErrorTree(self):
        node1 = TreeNode(1)
        node2 = TreeNode(2)
        node3 = TreeNode(3)
        node1.left = node2
        node2.left = node3

        self.flatten(node1)
        print('result')

    def test_NormalTree(self):
        node1 = TreeNode(1)
        node2 = TreeNode(2)
        node3 = TreeNode(3)
        node4 = TreeNode(4)
        node5 = TreeNode(5)
        node6 = TreeNode(6)
        node1.left = node2
        node1.right = node5
        node2.left = node3
        node2.right = node4
        node5.right = node6

        self.flatten(node1)
        print('result')

    def test_SingleNode(self):
        node1 = TreeNode(5)

        self.flatten(node1)
        print('result')

if __name__ == '__main__':
    unittest.main()